src/aggregation/agg_result.rs

```diff
@@ -10,25 +10,80 @@ use std::collections::HashMap;
 use itertools::Itertools;
 use serde::{Deserialize, Serialize};
 
+use super::agg_req::{Aggregations, AggregationsInternal, BucketAggregationInternal};
+use super::bucket::intermediate_buckets_to_final_buckets;
 use super::intermediate_agg_result::{
-    IntermediateAggregationResult, IntermediateAggregationResults, IntermediateBucketResult,
+    IntermediateAggregationResults, IntermediateBucketResult, IntermediateHistogramBucketEntry,
     IntermediateMetricResult, IntermediateRangeBucketEntry,
 };
 use super::metric::{SingleMetricResult, Stats};
 use super::Key;
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
 /// The final aggegation result.
 pub struct AggregationResults(pub HashMap<String, AggregationResult>);
 
-impl From<IntermediateAggregationResults> for AggregationResults {
-    fn from(tree: IntermediateAggregationResults) -> Self {
-        Self(
-            tree.0
-                .into_iter()
-                .map(|(key, agg)| (key, agg.into()))
-                .collect(),
-        )
+impl AggregationResults {
+    /// Convert and intermediate result and its aggregation request to the final result
+    pub fn from_intermediate_and_req(
+        results: IntermediateAggregationResults,
+        agg: Aggregations,
+    ) -> Self {
+        AggregationResults::from_intermediate_and_req_internal(results, &(agg.into()))
+    }
+    /// Convert and intermediate result and its aggregation request to the final result
+    ///
+    /// Internal function, CollectorAggregations is used instead Aggregations, which is optimized
+    /// for internal processing
+    fn from_intermediate_and_req_internal(
+        results: IntermediateAggregationResults,
+        req: &AggregationsInternal,
+    ) -> Self {
+        let mut result = HashMap::default();
+
+        // Important assumption:
+        // When the tree contains buckets/metric, we expect it to have all buckets/metrics from the
```

Comment:
What happens if this is not the case?
Does it panic, returns a wrong result or returns an error?

We want to avoid panic here, because IntermediateAggregationResults is coming from another node.